// react-templates/dist-test/src/shell.spec.js
'use strict';

var context = require('../../src/context');
var _ = require('lodash');
var path = require('path');

module.exports = {
    runTests: function runTests(test, dataPath) {
        test('test context', function (t) {
            context.clear();
            t.equal(context.hasErrors(), false);
            context.error('hi', '', 1, 1);
            t.equal(context.hasErrors(), true);
            context.clear();
            t.equal(context.hasErrors(), false);

            t.end();
        });

        test('test shell', function (t) {
            var shell = require('../../src/shell');
            var newContext = _.cloneDeep(context);
            var outputJSON = '';
            newContext.options.format = 'json';
            newContext.report = function (text) {
                outputJSON = text;
            };
            var r = shell.printResults(newContext);
            t.equal(r, 0);
            context.error('hi', '', 1, 1);
            r = shell.printResults(newContext);
            t.equal(r, 1);
            var output = JSON.parse(outputJSON);
            t.deepEqual(output, [{
                column: 1,
                endOffset: -1,
                file: null,
                index: -1,
                level: 'ERROR',
                line: 1,
                msg: 'hi',
                startOffset: -1
            }]);
            context.clear();
            t.end();
        });

        test('test shell', function (t) {
            var filename = path.join(dataPath, 'div.rt');
            var cli = require('../../src/cli');
            var r = cli.execute(filename + ' -r --dry-run');
            t.equal(r, 0);
            t.end();
        });
    }
};
